pkg/container/versioned/value.go

```diff
@@ -0,0 +1,337 @@
+	}
+	newNode.versions.past.Store(invalidVersion)
+
+	// find if there is a current value that is valid for this new version
+	node := v.head.Load()
+	prev := &v.head
+	for node != nil {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("AddValueAtVersion may not add values at versions lower than those already added! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
+		}
+
+		if node.versions.Contains(version) {
+			// truncate the validity of this node to end at 'version'
+			node.versions.past.Store(version)
+			// link the new node after the current one
+			newNode.next.Store(node.next.Load())
+			node.next.Store(newNode)
+			break
+		}
+
+		node = node.next.Load()
+		prev = &node.next
+	}
+	if node == nil {
+		// Add the new value at the end
+		prev.Store(newNode)
+	}
+}
+
+// RemoveValueAtVersion changes the validity of the stored value valid at 'version' to have ended at
+// 'version'.
+// 'version' must be later than any the current version visible to the readers.
+// Must be called from a single writer!
+func RemoveValueAtVersion[T any](v *Value[T], version Version) {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if version < node.versions.first {
+			stacktrace := hclog.Stacktrace()
+			log.Errorf("DeleteValueAtVersion may not be called with version lower than existing already! (%d < %d). Stacktrace: %s", version, node.versions.first, stacktrace)
+		}
+
+		if node.versions.Contains(version) {
+			// truncate the validity of this node to end at 'version'
+			node.versions.past.Store(version)
+			break
+		}
+	}
+}
+
+// Cleaner removes all values whose validity ends at or before 'version'.
+//
+// Must be called from a single writer!
+func Cleaner[T any](v *Value[T], keepVersion Version) {
+	// find all values that are no longer valid at 'version'
+	node := v.head.Load()
+	for node != nil && node.versions.past.Load() <= keepVersion {
+		// This node is no longer visible for readers who are all version 'keepVersion' or
+		// later.
+		node = node.next.Load()
+		v.head.Store(node)
+	}
+}
+
+// GetValue returns value of type 'T' valid for the given handle, or an empty value if none is
+// found.
+func GetValue[T any](v *Value[T], handle Handle) T {
+	for node := v.head.Load(); node != nil; node = node.next.Load() {
+		if node.versions.Contains(handle.getVersion()) {
+			return node.value
+		}
+	}
+	var empty T
+	return empty
+}
+
+// Pair is a struct of a version and any type T
+type Pair[T any] struct {
+	version Version
+	value   T
+}
+
+// PairSlice is a slice of Pairs
+type PairSlice[T any] []Pair[T]
+
+// NewPairSlice returns a new PairSlice of given 'capacity'
+func NewPairSlice[T any](capacity int) PairSlice[T] {
+	return make(PairSlice[T], 0, capacity)
+}
+
+// AppendPair appends a pair of 'version' and 'value' to PairSlice 's', returning 's'
+func AppendPair[T any](s PairSlice[T], version Version, value T) PairSlice[T] {
+	s = append(s, Pair[T]{
```

Comment:
just `return append(s, ...)`
and since then it's a single line function, is this really worth having?

Reply:
lets see...

Reply:
Removed unnecessary types related to this, but a function for this is needed so that the members can be private. Not sure it that is important, though.